### PR TITLE
[Gazebo 9] Created a MarkerArray to publish normals in ROS

### PR DESCRIPTION
**Pull request by ahcorde**
Moved this PR #1038 to `noetic-devel` to avoid ABI changes in `melodic`

We are exposing [normals in Gazebo9](https://bitbucket.org/osrf/gazebo/pull-requests/3193/added-normals-in-depth-camera-sensor/diff). This PR exposes the normals as a MarkerArray and then we can visualize the normals in RVIZ

Selección_014

Steps to reproduce it:

 - Launch the test world: 
`rosrun gazebo_ros gazebo test_worlds/gazebo_ros_depth_camera.world`
 - Create the tf between the world and the camera: 
`rosrun tf2_ros static_transform_publisher 0 0 5 0 0 -1.57 world base_link`
 - Open RVIZ

### CODE REVIEW

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_depth_camera.h
Comment (chapulina) [inline gazebo_plugins/include/gazebo_plugins/gazebo_ros_depth_camera.h:96]: Could you add `override`? I know the other functions don't have it, but it's probably because they predate C++11.

The doxygen also doesn't make sense to me. You could just write `// Documentation inherited` since the upstream function is already documented.

File: gazebo_plugins/include/gazebo_plugins/gazebo_ros_depth_camera.h
Comment (chapulina) [inline gazebo_plugins/include/gazebo_plugins/gazebo_ros_depth_camera.h:114]: Mind documenting all the new functions?

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_depth_camera.cpp]: Please remove the extra line at the end of the block, some linters complain about this

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_depth_camera.cpp]:
```suggestion
      if (pcd_ == nullptr)
```

Style nit :grin:

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_depth_camera.cpp]:
```suggestion
  if (this->point_cloud_msg_.data.size() > 0) {
```

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_depth_camera.cpp]:
```suggestion
        if (index % 50 == 0)
```

File: gazebo_plugins/test/test_worlds/gazebo_ros_depth_camera.world
Comment (chapulina) [inline gazebo_plugins/test/test_worlds/gazebo_ros_depth_camera.world]: Did you mean to change the path? Since the default is disabled, I think the whole `save` block could be removed.

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (iche033) [inline gazebo_plugins/src/gazebo_ros_depth_camera.cpp:290]: delete this in the destructor?

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (iche033) [inline gazebo_plugins/src/gazebo_ros_depth_camera.cpp]: hmm if there are still subscribers to the point cloud topic (or other topics advertised by the depth camera), this line will disable the sensor causing no msgs to be published. I think we should disable the sensor only when there are no subscribers to all the topics

**Review by chapulina — COMMENTED** [review]
I haven't looked into why, but when I `rosrun` the commands on the PR description, Gazebo just hangs with a black screen. Can you reproduce it? If you can't, I can dig into my setup later. The `worlds/depth_camera2.world` world doesn't hang though.

File: gazebo_plugins/test/test_worlds/gazebo_ros_depth_camera.world
Comment (iche033) [inline gazebo_plugins/test/test_worlds/gazebo_ros_depth_camera.world]: we can remove the <save> tag as it's not needed

**Comment by ahcorde** [comment]
You should run the example with `world/gazebo_ros_depth_camera.world` inside this package.

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_depth_camera.cpp]: Wouldn't it be best if this was controlled by the camera's frame rate? Right now it's hidden from the end user and they can't change it.

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (chapulina) [inline gazebo_plugins/src/gazebo_ros_depth_camera.cpp:391]: There's currently an implicit dependency between `points` and `normals`. If I run the test world, then echo only the normals, I get empty marker arrays. Once I also echo the points, then the marker array is populated.

I understand that  in order to create the markers we need the XYZ coordinates as well as the normals. So I think that on `OnNewRGBPointCloud`, we should also populate `point_cloud_msg` in case `this->normals_connect_count_ > 0`.

**Review by chapulina — CHANGES_REQUESTED** [review]

**Review by chapulina — DISMISSED** [review]
Works for me. Since @iche033 is also reviewing, let's wait for his approval before merging.

**Comment by chapulina** [comment]
Actually, I just realized that this change will make this branch not compilable against earlier Gazebo 9 versions. @j-rivero , what's the minimum Gazebo 9 version that we need to support for Noetic? I imagine that's 9.12.0, which is going into Focal? In that case, we may need to wrap the new APIs with `ifdef`s.

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (iche033) [inline gazebo_plugins/src/gazebo_ros_depth_camera.cpp:398]: move this check earlier? We could skip creating the marker msg and doing computations when the results are not being published.

**Comment by iche033** [comment]
looks good to me! just left one minor comment.

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (ahcorde) [reply gazebo_plugins/src/gazebo_ros_depth_camera.cpp:398]: fixed ;)

**Review by chapulina — APPROVED** [review]
> compiling against 9.12.0

I talked to @j-rivero about this and since the official Gazebo version for Noetic is not decided yet, let's not worry about it now. We can go back and fix it if needed.

**Review by iche033 — APPROVED** [review]

File: gazebo_plugins/src/gazebo_ros_depth_camera.cpp
Comment (iche033) [reply gazebo_plugins/src/gazebo_ros_depth_camera.cpp:398]: sorry didn't catch this earlier: `index` is declared later so it is not resolved here and this `if` block is missing curly brackects

Comment (ahcorde) [reply gazebo_plugins/src/gazebo_ros_depth_camera.cpp:398]: I had this resolved in this other PR, but I can create a Pr just solving this issue

**Pull request merged**
